kore/test/Test/ConsistentKore.hs

```diff
@@ -0,0 +1,849 @@
+import qualified Data.List as List
+    ( foldl'
+    )
+import qualified Data.Map as Map
+import Data.Maybe
+    ( catMaybes
+    , fromMaybe
+    )
+import qualified Data.Set as Set
+import Data.Text
+    ( Text
+    )
+import qualified Data.Text as Text
+
+import qualified Kore.Attribute.Symbol as Attribute
+    ( Symbol
+    )
+import qualified Kore.Builtin.Bool.Bool as BuiltinBool
+    ( asBuiltin
+    )
+import qualified Kore.Builtin.Int.Int as BuiltinInt
+    ( asBuiltin
+    )
+import qualified Kore.Builtin.List.List as BuiltinList
+    ( asBuiltin
+    )
+import qualified Kore.Builtin.String.String as BuiltinString
+    ( asBuiltin
+    )
+import qualified Kore.Domain.Builtin as Domain
+import Kore.IndexedModule.MetadataTools
+    ( SmtMetadataTools
+    )
+import qualified Kore.Internal.Alias as Internal
+    ( Alias (Alias)
+    )
+import qualified Kore.Internal.Alias as Alias.DoNotUse
+import Kore.Internal.ApplicationSorts
+    ( ApplicationSorts (ApplicationSorts)
+    )
+import qualified Kore.Internal.Symbol as Internal
+    ( Symbol (Symbol)
+    )
+import qualified Kore.Internal.Symbol as Symbol.DoNotUse
+import Kore.Internal.TermLike
+    ( TermLike
+    , TermLikeF (..)
+    , mkAnd
+    , mkApplyAlias
+    , mkApplySymbol
+    , mkBottom
+    , mkBuiltin
+    , mkCeil
+    , mkDomainValue
+    , mkElemVar
+    , mkEquals
+    , mkEvaluated
+    , mkExists
+    , mkFloor
+    , mkForall
+    , mkIff
+    , mkImplies
+    , mkIn
+    , mkMu
+    , mkNot
+    , mkNu
+    , mkOr
+    , mkRewrites
+    , mkSetVar
+    , mkStringLiteral
+    , mkTop
+    )
+import Kore.Sort
+    ( Sort
+    )
+import Kore.Syntax.DomainValue
+    ( DomainValue (DomainValue)
+    )
+import qualified Kore.Syntax.DomainValue as DomainValue.DoNotUse
+import Kore.Syntax.ElementVariable
+    ( ElementVariable (ElementVariable)
+    )
+import Kore.Syntax.SetVariable
+    ( SetVariable (SetVariable)
+    )
+import Kore.Syntax.Variable
+    ( Concrete
+    , Variable (Variable)
+    )
+import qualified Kore.Syntax.Variable as Variable.DoNotUse
+import Kore.Variables.UnifiedVariable
+    ( UnifiedVariable (..)
+    )
+
+import Test.Kore
+    ( idGen
+    )
+
+data SortRequirements
+    = AnySort
+    | SpecificSort !Sort
+    deriving (Eq, Ord, Show)
+
+data TermGenerator = TermGenerator
+    { arity :: !Integer
+    , sort :: !SortRequirements
+    , generator
+        :: (Sort -> Gen (TermLike Variable))
+        -> Sort
+        -> Gen (TermLike Variable)
+    }
+
+termGeneratorSort :: TermGenerator -> SortRequirements
+termGeneratorSort TermGenerator {sort} = sort
+
+data BuiltinGenerator = BuiltinGenerator
+    { arity :: !Integer
+    , sort :: !SortRequirements
+    , generator
+        :: (Sort -> Gen (TermLike Variable))
+        -> Gen (Domain.Builtin (TermLike Concrete) (TermLike Variable))
+    }
+
+data CollectionSorts = CollectionSorts
+    { collectionSort :: !Sort
+    , elementSort :: !Sort
+    }
+
+data Context = Context
```

Comment:
I think it would make sense to split this up: there are some fields which seem only to be used to prepare the `TermGenerator`s, and some which are only used during the actual generation (such as `allowedElementVariables`). That is, there seems to be a stage distinction here.

If you add a field like `currentSort :: Sort` alongside `allowedElementVariables`, then you can use `Gen.recursive` and use `Reader.local` to set the sort of the child pattern. However, I don't think you need to do this, I'm just remarking on how (I think) `Gen.recursive` is intended to be used.. The way it's written now seems perfectly natural to me.

Reply:
I split this, and didn't use `recursive`.